File: src/main/java/training/P010ClimbingStairs.java
package training;

public class P010ClimbingStairs {
    // Time: O(n), Space: O(1)
    public static int solve(int n) {
        if (n <= 2) {
            return n;
        }
        int prev2 = 1;
        int prev1 = 2;
        for (int i = 3; i <= n; i++) {
            int next = prev1 + prev2;
            prev2 = prev1;
            prev1 = next;
        }
        return prev1;
    }

    public static void main(String[] args) {
        System.out.println(solve(2));
        System.out.println(solve(5));
    }
}

File: training/python/p010_climbing_stairs.py
"""P010 Climbing Stairs
Time: O(n), Space: O(1)
"""

def solve(n):
    if n <= 2:
        return n
    prev2, prev1 = 1, 2
    for _ in range(3, n + 1):
        prev2, prev1 = prev1, prev1 + prev2
    return prev1


if __name__ == '__main__':
    print(solve(2))
    print(solve(5))
